Improve behavior of namespace flags in operator-sdk run and operator-sdk test local subcommands

nikhil-thomas opened the pull request:
Ref: Rebasing  **Improve flag names and behavior in up local and test sub-commands** #2149
as the code has diverged a lot

Fixes https://github.com/operator-framework/operator-sdk/issues/2644

# Summary
Rename `--namespace` flag as `--watch-namespace` and use only
with `operator-sdk run `

Add a `--operator-namespace` flag to be used only with `operator-sdk run --olm`

Rename `--namespace` flag in `operator-sdk test local --up-local` as `--operator-namespace`.

Add `--watch-namespace` flag in `operator-sdk test local --up-local` to explicitly specify watchNamespace (even `--watch-namespace=""` to watch all namespace)

---

This patch enables an operator to be tested locally with explicitly setting
`WATCH_NAMESPACE=""`

This PR implements the behavior discussed and decided in the pr,
[2100 Allow WATCH_NAMESPACE="" in operator-sdk test with --up-local flag
](https://github.com/operator-framework/operator-sdk/pull/2100)

Hence, the operator under test (localRunMode) can watch primary/secondary resource events from any namespace.

---

# Description of the change:**

## In operator-sdk run command
1. Change `--namespace` to  `--watch-namespace` in `operator-sdk run`, preserve
1. Add a `--operator-namespace` flag to be used only with `operator-sdk run --olm`

## In operator-sdk test local --up-local command
1. Change `--namespace` to `--operator-namespace` in `operator-sdk test local` (`--operator-namespace`: The operator where namespaced manifests are created)
1. Add `--watch-namespace` to `operator-sdk test local` (`--watch-namespace`: works only with `--up-local`, the namespace where the operator watches for changes in `operator-sdk test local`)

---

# Behavior

- `--operator-namespace` flag (in `operator-sdk test local --up-local`) specifies namespace where namespaced manifests are deployed for testing

`--operator-namespace` flag (in `operator-sdk test local --up-local`) preserves the behavior of old `--namespace` flag in operator-sdk test local`

- `--watch-namespace` flag (in `operator-sdk test local --up-local`) specifies the namespace which operator watches. That is the value of `WATCH_NAMESPACE` envvar

## `--watch-namespace` flag in operator-sdk test local

- `--watch-namespace=foo` 
    `WATCH_NAMESPACES=foo` operator watches namespace `foo` during testing

- `--watch-namespace=""` 
    `WATCH_NAMESPACES=""` operator watches all namespaces during testing

- when not set 
    `WATCH_NAMESPACE` is set to operator-namespace. 
    which means when `--watch-namespace` is not set, watch-namespace == operator-namespace
    operator watches deploy-namespace during testing

## `--watch-namespace` flag in operator-sdk run

already existing behavior of old `--namespace` flag is preserved in `--watch-namespace` flag

- `--watch-namespace=foo` 
    `WATCH_NAMESPACES=foo` operator watches namespace `foo`

- `--watch-namespace=""` 
    `WATCH_NAMESPACES=""` operator watches all namespaces

- when not set 
    `WATCH_NAMESPACE` is set to default namespace from kubeconfig

<!--

Before making a PR, please read our contributing guidelines https://github.com/operator-framework/operator-sdk/blob/master/CONTRIBUTING.MD

Note: Make sure your branch is rebased to the latest upstream master.

-->

---

# Motivation for the change

We couldn't receive events on secondary resources when they created in a namespace other than the namespace specified by `--namespace` flag, when testing with `--up-local` flag. This is not a problem while deploying the operator or testing without `--up-local` flag as the `WATCH_NAMESPACE=""` set in operator.yaml deployment manifest.

motivated by this patch is https://github.com/tektoncd/operator/issues/34.
Fixes: https://github.com/operator-framework/operator-sdk/issues/2644
Signed-off-by: Nikhil Thomas <nikthoma@redhat.com>

<!--

Before making a PR, please read our contributing guidelines https://github.com/operator-framework/operator-sdk/blob/master/CONTRIBUTING.MD

Note: Make sure your branch is rebased to the latest upstream master.

-->

**Description of the change:**


**Motivation for the change:**

<!--

Note: If this PR is fixing an issue make sure to add a note saying:
Closes #<ISSUE_NUMBER>

-->

camilamacedo86 commented on cmd/operator-sdk/test/local.go:
```diff
@@ -73,7 +73,7 @@ func newTestLocalCmd() *cobra.Command {
 	testCmd.Flags().StringVar(&tlConfig.moleculeTestFlags, "molecule-test-flags", "",
 		"Additional flags to pass to molecule test")
-	testCmd.Flags().StringVar(&tlConfig.namespace, "namespace", "",
+	testCmd.Flags().StringVar(&tlConfig.operatorNamespace, "operator-namespace", "",
```
We need an entry in the changelog with. See the other ones which were commands breaking changes.

nikhil-thomas commented:
@camilamacedo86 I have updated my patch. could you take a look

nikhil-thomas commented:
@joelanford 
> In the case where we're not in up-local mode and watch-namespace is not explicitly set, this will set TestWatchNamespaceEnv to the empty string, which is the equivalent of watch all namespaces.

>I guess since we're not in up-local mode this environment variable would just get ignored, but I can see how that would be confusing.

>Should we only set TestWatchNamespaceEnv if up-local mode is enabled?
(ref: https://github.com/operator-framework/operator-sdk/pull/2149#discussion_r349138337)

I have removed the `-singleNamespace` flag from `pkg/test` Framework.

I replaced it with a mechanism which checks whether the `TEST_OPERATOR_NAMESPACE` and `TEST_WATCH_NAMESPACE` was set, using `os.LookupEnv() string, bool` (instead of `os.GetEnv() string), to decide which Namespace to use to deploy namespaced manifests.

-  https://github.com/nikhil-thomas/operator-sdk/blob/3f9aab1341e6742aa6d0c64a7a37e4a23d85f6bc/pkg/test/framework.go#L121

- https://github.com/nikhil-thomas/operator-sdk/blob/3f9aab1341e6742aa6d0c64a7a37e4a23d85f6bc/pkg/test/context.go#L62

- https://github.com/nikhil-thomas/operator-sdk/blob/3f9aab1341e6742aa6d0c64a7a37e4a23d85f6bc/pkg/test/resource_creator.go#L60

nikhil-thomas commented:
@camilamacedo86 
im getting a golang lint error in sanity check
> Running golangci-lint
cmd/operator-sdk/test/local.go:78: line is 173 characters (lll)
		"Namespace where the operator will be deployed, CRs will be created and tests will be executed (By default it will be in the default namespace defined in the kubeconfig)")

> cmd/operator-sdk/test/local.go:80: line is 175 characters (lll)
		"(only valid with --up-local) Namespace where the operator watches for changes. Explicitly set to empty string to watch all namespaces (defaults to the operatorNamespace).")

It looks like im exceeding the number of characters allowed for the flag usage description

What is the allowed length

nikhil-thomas commented:
@AlexNPavel @estroz

jmrodri commented:
/lgtm

camilamacedo86 commented on CHANGELOG.md:
```diff
@@ -29,6 +29,9 @@
 - Upgrade the Helm dependency version from `v3.0.1` to `v3.0.2`. ([#2621](https://github.com/operator-framework/operator-sdk/pull/2621)) 
 - Changed the scaffolded `serveCRMetrics` to use the namespaces informed in the environment variable `WATCH_NAMESPACE` in the MultiNamespace scenario. ([#2603](https://github.com/operator-framework/operator-sdk/pull/2603)) 
 - Improve skip metrics logs when running the operator locally in order to make clear the information for Helm based operators. ([#2603](https://github.com/operator-framework/operator-sdk/pull/2603))
+- Breaking Change: Renamed --namespace flag as --watch-namespace in operator-sdk run local command. (#2617)
```
```suggestion
- **Breaking Change:** The `--namespace` flag was renamed for `--watch-namespace` in the `operator-sdk run ----local` command. ([#2617](https://github.com/operator-framework/operator-sdk/pull/2617))
```

camilamacedo86 commented on CHANGELOG.md:
```diff
@@ -29,6 +29,9 @@
+- Breaking Change: Renamed --namespace flag as --watch-namespace in operator-sdk run local command. (#2617)
+- Breaking Change: Split --namespace flag into --operator-namespace and --watch-namespace in operator-sdk test local command to decouple the operator namespace from the watch namespace. (#2617)
+- Breaking Change: Split `ctx.GetNamespace()` in `pkg/test` into `ctx.GetOperatorNamespace()` and `ctx.GetWatchNamespace()`. (#2617) 
 
```
In the "Deprecated" section:

```suggestion
- **Breaking Change:** The `--namespace` flag from `operator-sdk run --local` and `operator-sdk test --local` command was deprecated and will be removed in the future versions. Use `--watch-namespace` and `--operator-namespace`  instead of. ([#2617](https://github.com/operator-framework/operator-sdk/pull/2617))
- **Breaking Change:** The method `ctx.GetNamespace()` from the `pkg/test` is deprecated and will be removed in the future versions. Use `ctx.GetOperatorNamespace()` and `ctx.GetWatchNamespace()` instead of. ([#2617](https://github.com/operator-framework/operator-sdk/pull/2617))
```

In the "Added" section:

``
- The flag `--watch-namespace` and `--operator-namespace` was added to `operator-sdk run --local` and `operator-sdk test --local` commands in order to replace the flag `--namespace` which was  deprecated.([#2617](https://github.com/operator-framework/operator-sdk/pull/2617))
- The methods `ctx.GetOperatorNamespace()` and `ctx.GetWatchNamespace()` was added `pkg/test` in order to replace `ctx.GetNamespace()` which is  deprecated. ([#2617](https://github.com/operator-framework/operator-sdk/pull/2617))
```

NOTES:
- Use the format for `Breaking Change:` that has been used `**Breaking Change:**`
- Use the format to add the pr which has been used
- We should deprecate the current flags and methods instead of remove.

camilamacedo86 commented on cmd/operator-sdk/run/cmd.go:
```diff
@@ -33,7 +33,6 @@ import (
 type runCmd struct {
 	// Common options.
 	kubeconfig string
-	namespace  string
```
IHMO; we should deprecate and then remove in the next version.

nikhil-thomas commented:
> * The changelog entries need to follow the same standard convection done on it
i shall update those.

> * I think we should deprecate and then remove the  flags, methods in the next version
make sense :+1: , I shall attempt that. I don't think it will cause any issues

> * If be decided not deprecated and do it directly then, the changelog entries section in POV would be removed and added section.
i shall do this if the deprecation is not possible

> * Could you please add the break changes commands in the migration doc? See: https://github.com/operator-framework/operator-sdk/blob/master/doc/migration/version-upgrade-guide.md#unreleased. It would be like was done [here](https://github.com/operator-framework/operator-sdk/blob/master/doc/migration/version-upgrade-guide.md#breaking-changes-on-commands)
i shall

camilamacedo86 commented on pkg/test/resource_creator_test.go:
```diff
@@ -81,108 +83,115 @@ func TestGetWatchNamespace(t *testing.T) {
+			if len(operatorNamespace) <= 0 {
+				t.Errorf("expected non-empty operatorNamespace")
+			}
+			if watchNamespace != operatorNamespace {
```
Why the watchNamespace should be == operatorNamespace?

camilamacedo86 commented on pkg/test/resource_creator_test.go:
```diff
@@ -33,15 +33,16 @@ func TestGetOperatorNamespace(t *testing.T) {
-		}
-	})
+	t.Run("should create a non-empty new Operator Namespace when OperatorNamespace env is not set",
+		func(t *testing.T) {
```
Just change the indentation? 👍

nikhil-thomas replied:
@camilamacedo86 the `func` is intented like this to preserve the `80 character` lint rule.
i have kept the intendation consistent in other test cases.

nikhil-thomas replied:
the indendation is automatically done by `go fmt`

camilamacedo86 commented on pkg/test/resource_creator_test.go:
```diff
@@ -58,20 +59,21 @@ func TestGetOperatorNamespace(t *testing.T) {
+				Global.OperatorNamespace = ""
+			}()
+			ctx := NewContext(t)
+			//defer os.Unsetenv(OperatorNamespaceEnv)
```
```suggestion
```

nikhil-thomas replied:
done

camilamacedo86 commented on pkg/test/resource_creator_test.go:
```diff
@@ -58,20 +59,21 @@ func TestGetOperatorNamespace(t *testing.T) {
-		}
-	})
+	t.Run("should return non-empty new Operator Namespace, when OperatorNamespace Env = \"\"",
+		func(t *testing.T) {
```
Jus change the indentation 👍

nikhil-thomas replied:
@camilamacedo86 the func is intented like this to preserve the 80 character lint rule.
i have kept the intendation consistent in other test cases.

nikhil-thomas replied:
the indendation is automatically done by `go fmt`

camilamacedo86 commented on pkg/test/resource_creator_test.go:
```diff
@@ -81,108 +83,115 @@ func TestGetWatchNamespace(t *testing.T) {
+			if gotOperatorNamespace != operatorNamespace {
+				t.Errorf("expected operatorNamespace: %s, got %s", operatorNamespace, gotOperatorNamespace)
+			}
+		})
```
Could you confirm that you just change the indentation here?

nikhil-thomas replied:
@camilamacedo86 the func is intented like this to preserve the 80 character lint rule.
i have kept the intendation consistent in other test cases.

nikhil-thomas replied:
the indendation is automatically done by `go fmt`

camilamacedo86 commented on cmd/operator-sdk/run/cmd.go:
```diff
@@ -74,16 +73,24 @@ func NewCmd() *cobra.Command {
+				// set --watch-namespace flag if the --namespace flag is set
+				// (only if --watch-namespace flag is not set)
+				if cmd.Flags().Changed("namespace") {
+					log.Info("--namespace is deprecated; use --watch-namespace instead.")
```
👍 
```suggestion
					log.Info("--namespace is deprecated; use --watch-namespace and/or --operator-namespace instead.")
```

nikhil-thomas replied:
in `operator-sdk run` command the equivalent flag for old `--namespace` is `--watch-namespace`.
There is an already existing flag `--operator-namespace` in `operator-sdk run --local` command.

In fact. it is this `--operator-namespace` flag which made me adopt the name `--operator-namespace` in the `operator-sdk test --up-local` (instead of `--deploy-namespace` we planned in the initial discussions.

camilamacedo86 commented on cmd/operator-sdk/run/local.go:
```diff
@@ -51,6 +52,9 @@ type runLocalArgs struct {
 	prefix := "[local only] "
+
+	fs.StringVar(&c.watchNamespace, "watch-namespace", "",
+		prefix+"The namespace where the operator watches for changes.")
```
```suggestion
		prefix+"The namespace where the operator watches for changes. Note that the watch-namespace can be empty or a List of namespaces such as "nsa,nsb" for cluster scoped projects. However, the default matrics implementation requires that it contains the namespace where the operator is `deployed` in. ")
```

he

camilamacedo86 replied:
Feel free to re-phase ^

camilamacedo86 replied:
HI @nikhil-thomas. 

I'd like to clarify in the description. Note that now(master) the default implementation scaffolded is handling multi-namespaces. See:  https://github.com/camilamacedo86/operator-sdk/blob/master/internal/scaffold/cmd.go#L142

camilamacedo86 commented on doc/cli/operator-sdk_test_local.md:
```diff
@@ -21,11 +21,12 @@ operator-sdk test local <path to tests directory> [flags]
-      --namespace string              If non-empty, single namespace to run tests in
       --namespaced-manifest string    Path to manifest for per-test, namespaced resources (e.g. RBAC and Operator manifest)
       --no-setup                      Disable test resource creation
+      --operator-namespace string     Namespace where the operator will be deployed, CRs will be created and tests will be executed (By default it will be in the default namespace defined in the kubeconfig)
```
```suggestion
      --operator-namespace string     Namespace where the operator will be deployed (By default it will be in the default namespace defined in the kubeconfig)
```

camilamacedo86 commented on doc/cli/operator-sdk_test_local.md:
```diff
@@ -21,11 +21,12 @@ operator-sdk test local <path to tests directory> [flags]
+      --operator-namespace string     Namespace where the operator will be deployed, CRs will be created and tests will be executed (By default it will be in the default namespace defined in the kubeconfig)
       --skip-cleanup-error            If set as true, the cleanup function responsible to remove all artifacts will be skipped if an error is faced.
       --up-local                      Enable running operator locally with go run instead of as an image in the cluster
+      --watch-namespace string        (only valid with --up-local) Namespace where the operator watches for changes. Explicitly set to empty string to watch all namespaces (defaults to the operatorNamespace).
```
We have not more up-local. Also, is not it valid for the test as well?

nikhil-thomas replied:
@joelanford in a comment asked us to restrict it to `--up-local`.
if `--up-local` is not there then the `WATCH_NAMESPACE` is specified as a environment variable in `deploy/operator.yaml`

ref: https://github.com/operator-framework/operator-sdk/pull/2149#discussion_r349138337

camilamacedo86 replied:
Up local no longer exist. We have now operator-sdk run --local.

camilamacedo86 commented on doc/sdk-cli-reference.md:
```diff
@@ -493,6 +493,8 @@ Runs the tests locally
 * `--local-operator-flags` string - flags that the operator needs, while using --up-local (e.g. \"--flag1 value1 --flag2=value2\")
 * `--no-setup` - disable test resource creation
 * `--image` string - use a different operator image from the one specified in the namespaced manifest
+* --operator-namespace string     Namespace where the operator will be deployed, CRs will be created and tests will be executed (By default it will be in the default namespace defined in the kubeconfig)
```
```suggestion
* --operator-namespace string     Namespace where the operator will be deployed (By default it will be in the default namespace defined in the kubeconfig)
```

camilamacedo86 commented on doc/sdk-cli-reference.md:
```diff
@@ -493,6 +493,8 @@ Runs the tests locally
 * `--no-setup` - disable test resource creation
 * `--image` string - use a different operator image from the one specified in the namespaced manifest
+* --operator-namespace string     Namespace where the operator will be deployed, CRs will be created and tests will be executed (By default it will be in the default namespace defined in the kubeconfig)
+* --watch-namespace string        (only valid with --up-local) Namespace where the operator watches for changes. Explicitly set to empty string to watch all namespaces (defaults to the operatorNamespace).
```
dito  above

camilamacedo86 replied:
We need clarifies that it can be a List as well such as nsa,nsb,nsc

camilamacedo86 replied:
```suggestion
* --watch-namespace string        (only valid with --up-local) Namespace(s) where the operator watches for changes. Explicitly set to empty string to watch all namespaces (The default value is the Operator Namespace).
```

camilamacedo86 replied:
@nikhil-thomas ^

camilamacedo86 commented on cmd/operator-sdk/test/local.go:
```diff
@@ -74,7 +76,13 @@ func newTestLocalCmd() *cobra.Command {
+			"(By default it will be in the default namespace defined in the kubeconfig)")
+	testCmd.Flags().StringVar(&tlConfig.watchNamespace, "watch-namespace", "",
+		"(only valid with --up-local) Namespace where the operator watches for changes. "+
+			"Explicitly set to empty string to watch all namespaces (defaults to the operatorNamespace).")
```
- --up-local no longer exists
- see the above explanation over whatch_namespaces

nikhil-thomas replied:
@camilamacedo86 `--up-local` is there in `operator-sdk test local`, right.
It doesn't exist in `operator-sdk run`

camilamacedo86 replied:
You are right.  Here is the test command. So, the command is ` operator-sdk test local ./test/e2e --verbose` .

camilamacedo86 commented on cmd/operator-sdk/test/local.go:
```diff
@@ -91,6 +99,15 @@ func newTestLocalCmd() *cobra.Command {
+	// set --operator-namespace flag if the --namespace flag is set
+	// (only if --operator-namespace flag is not set)
+	if cmd.Flags().Changed("namespace") {
+		log.Info("--namespace is deprecated; use --operator-namespace instead.")
```
^ like above

camilamacedo86 commented on cmd/operator-sdk/test/local.go:
```diff
@@ -130,8 +147,11 @@ func testLocalGoFunc(cmd *cobra.Command, args []string) error {
-	if tlConfig.upLocal && tlConfig.namespace == "" {
-		return fmt.Errorf("must specify a namespace to run in when -up-local flag is set")
+	if tlConfig.upLocal && tlConfig.operatorNamespace == "" {
+		return fmt.Errorf("must specify a namespace with operator-namespace flag to run in when --up-local flag is set")
```
Same ^

camilamacedo86 commented on cmd/operator-sdk/test/local.go:
```diff
@@ -130,8 +147,11 @@ func testLocalGoFunc(cmd *cobra.Command, args []string) error {
+		return fmt.Errorf("must specify a namespace with operator-namespace flag to run in when --up-local flag is set")
+	}
+	if !tlConfig.upLocal && cmd.Flags().Changed("watch-namespace") {
+		return fmt.Errorf("--watch-namespace not valid without -up-local flag")
```
Same

camilamacedo86 commented on cmd/operator-sdk/test/local.go:
```diff
@@ -220,10 +240,24 @@ func testLocalGoFunc(cmd *cobra.Command, args []string) error {
-	env := append(os.Environ(), fmt.Sprintf("%v=%v", test.TestNamespaceEnv, tlConfig.namespace))
+	env := os.Environ()
+	if tlConfig.operatorNamespace != "" {
+		env = append(
```
WDYT about add a log to inform that the it will be usin the operatorNamespace == value?

nikhil-thomas replied:
we cannot be sure about it here.
if the `operator-namespace` is not set explicitly, the a operaotor-namespace is created by
`ctx.GetOperatorNamespace()` (or in the deprecated `ctx.GEtNamespace()`).

It is created at random
`ctx.namespace = ctx.GetID()`

so a log here can be confusing.

camilamacedo86 commented on cmd/operator-sdk/test/local.go:
```diff
@@ -220,10 +240,24 @@ func testLocalGoFunc(cmd *cobra.Command, args []string) error {
+	}
+
+	if cmd.Flags().Changed("watch-namespace") {
+		env = append(
```
same here

nikhil-thomas replied:
https://github.com/operator-framework/operator-sdk/pull/2617#discussion_r391679380

In addition if both `operatorNamespace` and `watchNamespace` is not set.
operatorNamespace is created during  while calling xtx.GetOperatorNamepspace() and watchNamespace is set to operator namespace.

I have considered all of those in units tests in 
pkg/test/resource_creator_test.go

camilamacedo86 commented on doc/cli/operator-sdk_cleanup.md:
```diff
@@ -13,15 +13,16 @@ operator-sdk cleanup [flags]
+      --kubeconfig string           The file path to kubernetes configuration file. Defaults to location specified by $KUBECONFIG, or to default file rules if not set
+      --namespace string            The namespace where the operator watches for changes.
+      --olm                         The operator to be cleaned up is managed by OLM in a cluster. Cannot be set with another cleanup-type flag (default true)
+      --operator-namespace string   [olm only] The namespace where operator resources are created in --olm mode. It must already exist in the cluster or be defined in a manifest passed to IncludePaths.
```
same ^

camilamacedo86 replied:
```suggestion
      --operator-namespace string   [olm only] The namespace where the operator is deployed in. 
```

jmrodri commented:
I hate to be that person, but it failed the linter :( 
```
cmd/operator-sdk/run/local.go:44:2: `namespace` is unused (structcheck)
	namespace            string
	^
pkg/test/context_test.go:25:27: string `fakePath` has 3 occurrences, make it a constant (goconst)
	fakeNamespacedManPath := "fakePath"
	                         ^
Makefile:81: recipe for target 'lint' failed
```

nikhil-thomas commented:
> I hate to be that person, but it failed the linter :(
> 
> ```
> cmd/operator-sdk/run/local.go:44:2: `namespace` is unused (structcheck)
> 	namespace            string
> 	^
> pkg/test/context_test.go:25:27: string `fakePath` has 3 occurrences, make it a constant (goconst)
> 	fakeNamespacedManPath := "fakePath"
> 	                         ^
> Makefile:81: recipe for target 'lint' failed
> ```

fixed it :)

camilamacedo86 commented on cmd/operator-sdk/test/local.go:
```diff
@@ -74,7 +76,13 @@ func newTestLocalCmd() *cobra.Command {
-		"If non-empty, single namespace to run tests in")
+		"(Deprecated: use --operator-namespace instead) If non-empty, single namespace to run tests in")
+	testCmd.Flags().StringVar(&tlConfig.operatorNamespace, "operator-namespace", "",
+		"Namespace where the operator will be deployed, CRs will be created and tests will be executed "+
```
```suggestion
		"Namespace where the operator will be deployed"+
```


The CR's are not created by the operator in the default implementation. It needs to be applied manually.

nikhil-thomas replied:
@camilamacedo86 
in operator-sdk test local --up-local the CRs are programmatically  created by users who write tests right.
whereas  in operator-sdk run --local a user will have to add the CRDs, CRS, RBAC etc manually, right

camilamacedo86 commented on cmd/operator-sdk/test/local.go:
```diff
@@ -74,7 +76,13 @@ func newTestLocalCmd() *cobra.Command {
+			"(By default it will be in the default namespace defined in the kubeconfig)")
+	testCmd.Flags().StringVar(&tlConfig.watchNamespace, "watch-namespace", "",
+		"(only valid with --up-local) Namespace where the operator watches for changes. "+
+			"Explicitly set to empty string to watch all namespaces (defaults to the operatorNamespace).")
```
Note that the WATCH can be 1...N values or an empty string. Could we clarify it here as well?

nikhil-thomas replied:
done

camilamacedo86 commented on doc/cli/operator-sdk_cleanup.md:
```diff
@@ -13,15 +13,16 @@ operator-sdk cleanup [flags]
-      --timeout duration          [olm only] Time to wait for the command to complete before failing (default 2m0s)
-  -h, --help                      help for cleanup
+      --kubeconfig string           The file path to kubernetes configuration file. Defaults to location specified by $KUBECONFIG, or to default file rules if not set
+      --namespace string            The namespace where the operator watches for changes.
```
```suggestion
      --namespace string            (Deprecated) - The namespace where the operator watches for changes.
```

camilamacedo86 replied:
Did you not change this as well?

camilamacedo86 replied:
I think we need to change all to keep the same behaviour across the project. Otherwise is hard to know when use  1 or another. Is it make sense?

nikhil-thomas replied:
done

camilamacedo86 commented on doc/cli/operator-sdk_test_local.md:
```diff
@@ -21,11 +21,13 @@ operator-sdk test local <path to tests directory> [flags]
+      --operator-namespace string     Namespace where the operator will be deployed, CRs will be created and tests will be executed (By default it will be in the default namespace defined in the kubeconfig)
       --skip-cleanup-error            If set as true, the cleanup function responsible to remove all artifacts will be skipped if an error is faced.
       --up-local                      Enable running operator locally with go run instead of as an image in the cluster
+      --watch-namespace string        (only valid with --up-local) Namespace where the operator watches for changes. Explicitly set to empty string to watch all namespaces (defaults to the operatorNamespace).
```
```suggestion
      --watch-namespace string        (only valid with --up-local) Namespace(s) where the operator watches for changes. Explicitly set to empty string to watch all namespaces (The default value is the Operator Namespace).
```

nikhil-thomas replied:
done

camilamacedo86 commented on pkg/test/resource_creator.go:
```diff
@@ -29,6 +30,9 @@ import (
 	"k8s.io/apimachinery/pkg/util/wait"
 )
 
+// Deprecated: GetNamespace() is deprecated.
```
```suggestion
// TODO: remove before 1.0.0
// Deprecated: GetNamespace() is deprecated.
```

nikhil-thomas replied:
done

camilamacedo86 commented on pkg/test/resource_creator.go:
```diff
@@ -29,6 +30,9 @@ import (
 )
 
+// Deprecated: GetNamespace() is deprecated.
+// GetNamespace() will be removed in future versions
```
```suggestion
```

nikhil-thomas replied:
done

camilamacedo86 commented on pkg/test/resource_creator.go:
```diff
@@ -48,8 +52,48 @@ func (ctx *Context) GetNamespace() (string, error) {
 	return ctx.namespace, nil
 }
 
+func (ctx *Context) GetOperatorNamespace() (string, error) {
```
```suggestion
// GetOperatorNamespace will return an Operator Namespace, if the flag --operator-namespace not be used then it will returna randon value. 
func (ctx *Context) GetOperatorNamespace() (string, error) {
```

nikhil-thomas replied:
done

camilamacedo86 commented on pkg/test/resource_creator.go:
```diff
@@ -48,8 +52,48 @@ func (ctx *Context) GetNamespace() (string, error) {
+	return ctx.operatorNamespace, nil
+}
+
+func (ctx *Context) GetWatchNamespace() (string, error) {
```
```suggestion
// GetWatchNamespace will return the  namespaces to operator watch for changes, if the flag --watch-namespaced not be used than it will  return the Operator Namespace. 
func (ctx *Context) GetWatchNamespace() (string, error) {
```

nikhil-thomas replied:
done

camilamacedo86 commented on test/test-framework/go.mod:
```diff
@@ -5,10 +5,10 @@ go 1.13
 	k8s.io/client-go v12.0.0+incompatible
-	sigs.k8s.io/controller-runtime v0.4.0
+	sigs.k8s.io/controller-runtime v0.5.1
 )
```
We should not do these updates. See that the master branch is using the old versions
https://github.com/operator-framework/operator-sdk/blob/master/go.mod#L62

camilamacedo86 replied:
Revert the go.mod and go.sum here.

nikhil-thomas replied:
thanks for pointing this out :+1: 
removed

camilamacedo86 commented on pkg/test/context.go:
```diff
@@ -56,14 +61,24 @@ func (f *Framework) newContext(t *testing.T) *Context {
 		id:                 id,
 		t:                  t,
-		namespace:          namespace,
+		namespace:          operatorNamespace,
```
```suggestion
                 // the  namespace is deprecated
                 // todo: remove before 1.0.0 
		namespace:          operatorNamespace,
```

nikhil-thomas replied:
done

camilamacedo86 commented on go.mod:
```diff
@@ -59,7 +60,7 @@ require (
 	k8s.io/kubectl v0.0.0
 	k8s.io/kubernetes v1.16.2 // indirect
-	sigs.k8s.io/controller-runtime v0.4.0
+	sigs.k8s.io/controller-runtime v0.5.1
```
Why upgrade the versions?
It would need to be done in another PR.

nikhil-thomas replied:
reverted

camilamacedo86 commented on doc/cli/operator-sdk_run.md:
```diff
@@ -13,19 +13,21 @@ operator-sdk run [flags]
+      --include strings             [olm only] Path to Kubernetes resource manifests, ex. Role, Subscription. These supplement or override defaults generated by run/cleanup
+      --timeout duration            [olm only] Time to wait for the command to complete before failing (default 2m0s)
+      --local                       The operator will be run locally by building the operator binary with the ability to access a kubernetes cluster using a kubeconfig file. Cannot be set with another run-type flag.
+      --watch-namespace string      [local only] The namespace where the operator watches for changes.
```
```suggestion
      --watch-namespace string      [local only]  Namespace(s) where the operator watches for changes. Explicitly set to empty string to watch all namespaces (The default value is the Operator Namespace)
```

camilamacedo86 commented on cmd/operator-sdk/test/local.go:
```diff
@@ -74,7 +76,13 @@ func newTestLocalCmd() *cobra.Command {
+		"Namespace where the operator will be deployed, CRs will be created and tests will be executed "+
+			"(By default it will be in the default namespace defined in the kubeconfig)")
+	testCmd.Flags().StringVar(&tlConfig.watchNamespace, "watch-namespace", "",
+		"(only valid with --up-local) Namespace where the operator watches for changes. "+
```
```suggestion
		"(only valid with --up-local) Namespace(s) where the operator watches for changes. "+
```

camilamacedo86 commented on cmd/operator-sdk/test/local.go:
```diff
@@ -74,7 +76,13 @@ func newTestLocalCmd() *cobra.Command {
+			"(By default it will be in the default namespace defined in the kubeconfig)")
+	testCmd.Flags().StringVar(&tlConfig.watchNamespace, "watch-namespace", "",
+		"(only valid with --up-local) Namespace where the operator watches for changes. "+
+			"Explicitly set to empty string to watch all namespaces (defaults to the operatorNamespace).")
```
```suggestion
			"Explicitly set to empty string to watch all namespaces (The default value is the Operator Namespace)")
```

nikhil-thomas replied:
done

camilamacedo86 commented on doc/migration/version-upgrade-guide.md:
````diff
@@ -972,6 +972,22 @@ With:
 ```yaml 
 - name: {{your operator name which is the value of metadata.name in this file}}
 ```
+### Notable Changes
````
Here we need to add it before [Remove Ansible container sidecar](https://github.com/operator-framework/operator-sdk/blob/e113b6c3caf643a192f172c6bec60922c601e977/doc/migration/version-upgrade-guide.md#remove-ansible-container-sidecar) in the Breaking change section as follows. 

--- 

### Breaking changes

#### In the test-framwork

The methods `ctx.GetOperatorNamespace()` and `ctx.GetWatchNamespace()` was added `pkg/test` in order to replace `ctx.GetNamespace()` which is  deprecated. In this way replace the use of `ctx.GetNamespace()` in your project with `ctx.GetOperatorNamespace()`

#### Remove Ansible container sidecar
... 
----

Then, regards the commands we need to another section after the Breaking Change above as follows. See the example: https://github.com/operator-framework/operator-sdk/blob/master/doc/migration/version-upgrade-guide.md#breaking-changes-on-commands

### Breaking Changes on Commands 

This release contains breaking changes in some commands.

- The `--namespace` flag from `operator-sdk run --local` and `operator-sdk test --local` command was deprecated and was replaced by `--watch-namespace` and `--operator-namespace` .  

Note that `--watch-namespace` can be used to set the namespace(s) which the operator will watch for changes. It will set the environment variable `WATCH_NAMESPACE`.  Use explicitly an empty string to watch all namespaces or inform a List of namespaces such as "ns1,ns2" when the operator is [cluster-scoped](https://github.com/operator-framework/operator-sdk/blob/master/doc/operator-scope.md#cluster-scoped-operator-usage). If you use a List, then it needs contains the namespace where the operator is "deployed" in since the default metrics implementation will manage resources in the Operator's namespace. By default, it will be the Operator Namespace. 

Then, use the flag `--operator-namespace` to inform the namespace where the Operator will be "deployed" in and then, it will set the environment variable `OPERATOR_NAMESPACE`. If this value is not set, then it will be the namespace defined as default in the Kubeconfig. 
---

nikhil-thomas replied:
done

camilamacedo86 commented on CHANGELOG.md:
```diff
@@ -16,6 +16,8 @@
 - Add the --proxy-port flag to the operator-sdk scorecard command allowing users to override the default proxy port value (8889). ([2634](https://github.com/operator-framework/operator-sdk/pull/2634))
 - Add support for Metrics with MultiNamespace scenario. ([#2603](https://github.com/operator-framework/operator-sdk/pull/2603)) 
 - Add Prometheus metrics support to Helm-based operators. ([#2603](https://github.com/operator-framework/operator-sdk/pull/2603))
+- The flag `--watch-namespace` and `--operator-namespace` was added to `operator-sdk run --local` and `operator-sdk test --local` commands in order to replace the flag `--namespace` which was  deprecated.([#2617](https://github.com/operator-framework/operator-sdk/pull/2617))
```
```suggestion
- The flag `--watch-namespace` and `--operator-namespace` was added to `operator-sdk run --local`, `operator-sdk test --local` and `operator-sdk cleanup` commands in order to replace the flag `--namespace` which was  deprecated.([#2617](https://github.com/operator-framework/operator-sdk/pull/2617))
```

nikhil-thomas replied:
done

camilamacedo86 commented on CHANGELOG.md:
```diff
@@ -30,6 +32,8 @@
 - Upgrade the Helm dependency version from `v3.0.1` to `v3.0.2`. ([#2621](https://github.com/operator-framework/operator-sdk/pull/2621)) 
 - Changed the scaffolded `serveCRMetrics` to use the namespaces informed in the environment variable `WATCH_NAMESPACE` in the MultiNamespace scenario. ([#2603](https://github.com/operator-framework/operator-sdk/pull/2603)) 
 - Improve skip metrics logs when running the operator locally in order to make clear the information for Helm based operators. ([#2603](https://github.com/operator-framework/operator-sdk/pull/2603))
+- **Breaking Change:** The `--namespace` flag from `operator-sdk run --local` and `operator-sdk test --local` command was deprecated and will be removed in the future versions. Use `--watch-namespace` and `--operator-namespace`  instead of. ([#2617](https://github.com/operator-framework/operator-sdk/pull/2617))
```
```suggestion
- **Breaking Change:** The `--namespace` flag from `operator-sdk run --local`, `operator-sdk test --local` and `operator-sdk cleanup` command was deprecated and will be removed in the future versions. Use `--watch-namespace` and `--operator-namespace`  instead of. ([#2617](https://github.com/operator-framework/operator-sdk/pull/2617))
```

nikhil-thomas replied:
done

camilamacedo86 commented on CHANGELOG.md:
```diff
@@ -30,6 +32,8 @@
 - Improve skip metrics logs when running the operator locally in order to make clear the information for Helm based operators. ([#2603](https://github.com/operator-framework/operator-sdk/pull/2603))
+- **Breaking Change:** The `--namespace` flag from `operator-sdk run --local` and `operator-sdk test --local` command was deprecated and will be removed in the future versions. Use `--watch-namespace` and `--operator-namespace`  instead of. ([#2617](https://github.com/operator-framework/operator-sdk/pull/2617))
+- **Breaking Change:** The method `ctx.GetNamespace()` from the `pkg/test` is deprecated and will be removed in the future versions. Use `ctx.GetOperatorNamespace()` and `ctx.GetWatchNamespace()` instead of. ([#2617](https://github.com/operator-framework/operator-sdk/pull/2617))
 
```
The above breaking changes need to go to the Deprecated section. https://github.com/operator-framework/operator-sdk/blob/e113b6c3caf643a192f172c6bec60922c601e977/CHANGELOG.md#deprecated

nikhil-thomas replied:
done

camilamacedo86 commented on cmd/operator-sdk/test/local.go:
```diff
@@ -48,6 +48,8 @@ type testLocalConfig struct {
 	goTestFlags        string
 	moleculeTestFlags  string
 	namespace          string
```
```suggestion
        // TODO: remove before 1.0.0
        // Namespace is deprecated  
	namespace          string
```

nikhil-thomas replied:
done

camilamacedo86 commented on cmd/operator-sdk/test/local.go:
```diff
@@ -74,7 +76,13 @@ func newTestLocalCmd() *cobra.Command {
 	testCmd.Flags().StringVar(&tlConfig.moleculeTestFlags, "molecule-test-flags", "",
 		"Additional flags to pass to molecule test")
 	testCmd.Flags().StringVar(&tlConfig.namespace, "namespace", "",
```
```suggestion
        // TODO: remove before 1.0.0. Namespace is deprecated 
	testCmd.Flags().StringVar(&tlConfig.namespace, "namespace", "",
```

nikhil-thomas replied:
done

camilamacedo86 commented on cmd/operator-sdk/test/local.go:
```diff
@@ -91,6 +99,15 @@ func newTestLocalCmd() *cobra.Command {
 }
 
 func testLocalFunc(cmd *cobra.Command, args []string) error {
+	// set --operator-namespace flag if the --namespace flag is set
```
```suggestion
        //TODO: remove before 1.0.0
	// set --operator-namespace flag if the --namespace flag is set
```

nikhil-thomas replied:
done

camilamacedo86 commented on cmd/operator-sdk/run/cmd.go:
```diff
@@ -74,16 +82,24 @@ func NewCmd() *cobra.Command {
 					log.Fatalf("Failed to run operator using OLM: %v", err)
 				}
 			case c.local:
+				// set --watch-namespace flag if the --namespace flag is set
```
//TODO: remove namespace flag before 1.0.0

nikhil-thomas replied:
i have checked again and add the comment in all relevant locations.

nikhil-thomas commented:
@camilamacedo86 
- [x] See the comments over the description of the commands and some comments in the code (nits)
- [x] See the comment over the migration guide. We need to fix it. https://github.com/operator-framework/operator-sdk/pull/2617/files#r392133791
- [x] Missing perform the same change for operator-sdk cleanup. We need to have the same behaviour across the project to make things more intuitive and easier to be used. So, the flag namespace on it needs to be deprecated and we need to add the watch_namespace. then, it means that we need to add this info in the changelog, docs and guide as well. See: https://github.com/operator-framework/operator-sdk/pull/2617/files#r392103614
- [x] Also, we need to revert the changes in the go.mod and go.sum files.
- [ ] The PR needs pass in the CI. See the error: https://travis-ci.org/github/operator-framework/operator-sdk/jobs/661801834#L895 (to test locally you can run make lint)
- [x] PS.: Can you please simplify(less verbose) the first comment and add as motivation the issue: #2644 as well?

camilamacedo86 commented on cmd/operator-sdk/cleanup/cmd.go:
```diff
@@ -29,7 +29,8 @@ import (
 	// Common options.
 	kubeconfig string
-	namespace  string
+	// deprecated: use olmArgs.OperatorNamespace instead
```
missing todo remove

camilamacedo86 commented on doc/migration/version-upgrade-guide.md:
```diff
@@ -940,6 +940,10 @@ func serveCRMetrics(cfg *rest.Config, operatorNs string) error {
 
+#### In the test-framwork
+
+The methods ctx.GetOperatorNamespace() and ctx.GetWatchNamespace() was added pkg/test in order to replace ctx.GetNamespace() which is deprecated. In this way replace the use of ctx.GetNamespace() in your project with ctx.GetOperatorNamespace()
```
```suggestion
The methods `ctx.GetOperatorNamespace()` and `ctx.GetWatchNamespace()` was added pkg/test in order to replace `ctx.GetNamespace()` which is deprecated. In this way replace the use of `ctx.GetNamespace()` in your project with `ctx.GetOperatorNamespace()`
```

nikhil-thomas replied:
done

camilamacedo86 commented on doc/migration/version-upgrade-guide.md:
```diff
@@ -973,6 +977,33 @@ With:
+
+Then, use the flag --operator-namespace to inform the namespace where the Operator will be "deployed" in and then, it will set the environment variable OPERATOR_NAMESPACE. If this value is not set, then it will be the namespace defined as default in the Kubeconfig.
+
+### Notable Changes
```
Missing remove the `### Notable Changes`

nikhil-thomas replied:
done

camilamacedo86 commented on doc/migration/version-upgrade-guide.md:
```diff
@@ -973,6 +977,33 @@ With:
+
+This release contains breaking changes in some commands.
+
+- The --namespace flag from operator-sdk run --local and operator-sdk test --local command was deprecated and was replaced by --watch-namespace and --operator-namespace .
```
Missing add the cleanup command.

nikhil-thomas replied:
added

camilamacedo86 commented on doc/migration/version-upgrade-guide.md:
```diff
@@ -973,6 +977,31 @@ With:
+- `--watch-namespace` is added to `operator-sdk test --up-local` command. This flag specifies the namespace in where the operator watches
+   for changes. This flag can be set to "foobar", "" (watch-all-namespace) or left unset (when not set watch-namespace = operator-namespace)*[]:
+- The `ctx.GetNamespace()` in `pkg/test` is replaced with `ctx.GetOperatorNamespace()` and `ctx.GetWatchNamespace()` to accomodate the
+  changes in flags.
```
I think we need to remove line 990 to 1001.
It is already described above.

nikhil-thomas replied:
done

camilamacedo86 commented on doc/migration/version-upgrade-guide.md:
```diff
@@ -940,6 +940,10 @@ func serveCRMetrics(cfg *rest.Config, operatorNs string) error {
 
```
@nikhil-thomas, 

The release was made. Then, is required to rebase with the master and just your sections/content will be added at the and : 

```
## Unreleased (Master branch)

### Breaking changes

#### In the test-framwork

The methods `ctx.GetOperatorNamespace()` and `ctx.GetWatchNamespace()` was added pkg/test in order to replace `ctx.GetNamespace()` which is deprecated. In this way replace the use of `ctx.GetNamespace()` in your project with `ctx.GetOperatorNamespace()`

### Breaking Changes on Commands

This release contains breaking changes in some commands.

- The --namespace flag from `operator-sdk run --local` command, `operator-sdk test --local` command and `operator-sdk cleanup` command was deprecated and was replaced by --watch-namespace and --operator-namespace .

Note that --watch-namespace can be used to set the namespace(s) which the operator will watch for changes. It will set the environment variable WATCH_NAMESPACE. Use explicitly an empty string to watch all namespaces or inform a List of namespaces such as "ns1,ns2" when the operator is cluster-scoped. If you use a List, then it needs contains the namespace where the operator is "deployed" in since the default metrics implementation will manage resources in the Operator's namespace. By default, it will be the Operator Namespace.

Then, use the flag --operator-namespace to inform the namespace where the Operator will be "deployed" in and then, it will set the environment variable OPERATOR_NAMESPACE. If this value is not set, then it will be the namespace defined as default in the Kubeconfig.
``` 

Note that just the above is required. The lines https://github.com/operator-framework/operator-sdk/pull/2617/files#diff-59e439e5373edfb5478c1999335f4842R990-R1001 can be removed.

nikhil-thomas replied:
done

camilamacedo86 replied:
See that your content was NOT removed from 0.16x section yet : https://github.com/operator-framework/operator-sdk/blob/56364ef881f6be5c077f3d9f424ce9d4fc192fc6/doc/migration/version-upgrade-guide.md#breaking-changes-on-commands-1

camilamacedo86 replied:
https://github.com/operator-framework/operator-sdk/pull/2617/files#diff-59e439e5373edfb5478c1999335f4842R980

camilamacedo86 commented on doc/migration/version-upgrade-guide.md:
```diff
@@ -991,6 +990,25 @@ RUN ansible-galaxy collection install -r ${HOME}/requirements.yml \
+Note that --watch-namespace can be used to set the namespace(s) which the operator will watch for changes. It will set the environment variable WATCH_NAMESPACE. Use explicitly an empty string to watch all namespaces or inform a List of namespaces such as "ns1,ns2" when the operator is cluster-scoped. If you use a List, then it needs contains the namespace where the operator is "deployed" in since ...
+
+Then, use the flag --operator-namespace to inform the namespace where the Operator will be "deployed" in and then, it will set the environment variable OPERATOR_NAMESPACE. If this value is not set, then it will be the namespace defined as default in the Kubeconfig.
+
```
We need to do some improvement in the above text.
However, we could to do it as a follow-up pr. The most important in IMO is we have the info here to help us in the release process.

nikhil-thomas replied:
sure, :+1:

camilamacedo86 commented on doc/migration/version-upgrade-guide.md:
```diff
@@ -991,6 +990,25 @@ RUN ansible-galaxy collection install -r ${HOME}/requirements.yml \
+
+This release contains breaking changes in some commands.
+
+- The --namespace flag from `operator-sdk run --local` command, `operator-sdk test --local` command and `operator-sdk cleanup` command was deprecated and was replaced by --watch-namespace and --operator-namespace .
```
```suggestion
- The `--namespace` flag from `operator-sdk run --local` command, `operator-sdk test --local` command and `operator-sdk cleanup` command was deprecated and was replaced by `--watch-namespace` and `--operator-namespace` .
```

camilamacedo86 commented on doc/migration/version-upgrade-guide.md:
```diff
@@ -991,6 +990,25 @@ RUN ansible-galaxy collection install -r ${HOME}/requirements.yml \
+
+- The --namespace flag from `operator-sdk run --local` command, `operator-sdk test --local` command and `operator-sdk cleanup` command was deprecated and was replaced by --watch-namespace and --operator-namespace .
+
+Note that --watch-namespace can be used to set the namespace(s) which the operator will watch for changes. It will set the environment variable WATCH_NAMESPACE. Use explicitly an empty string to watch all namespaces or inform a List of namespaces such as "ns1,ns2" when the operator is cluster-scoped. If you use a List, then it needs contains the namespace where the operator is "deployed" in since ...
```
```suggestion
Note that `--watch-namespace` flag can be used to set the namespace(s) which the operator will watch for changes. It will set the environment variable `WATCH_NAMESPACE`. Use explicitly an empty string to watch all namespaces or inform a List of namespaces such as; "ns1,ns2" when the operator is cluster-scoped. If you use a List, then it needs contains the namespace where the operator is "deployed" in since the default metrics implementation will manage resources in the Operator's namespace. By default, it will be the `Operator Namespace`.
```

camilamacedo86 commented on doc/migration/version-upgrade-guide.md:
```diff
@@ -991,6 +990,25 @@ RUN ansible-galaxy collection install -r ${HOME}/requirements.yml \
+
+Note that --watch-namespace can be used to set the namespace(s) which the operator will watch for changes. It will set the environment variable WATCH_NAMESPACE. Use explicitly an empty string to watch all namespaces or inform a List of namespaces such as "ns1,ns2" when the operator is cluster-scoped. If you use a List, then it needs contains the namespace where the operator is "deployed" in since ...
+
+Then, use the flag --operator-namespace to inform the namespace where the Operator will be "deployed" in and then, it will set the environment variable OPERATOR_NAMESPACE. If this value is not set, then it will be the namespace defined as default in the Kubeconfig.
```
```suggestion
Then, use the flag `--operator-namespace` to inform the namespace where the Operator will be "deployed" in and then, it will set the environment variable `OPERATOR_NAMESPACE`. If this value is not set, then it will be the namespace defined as default in the Kubeconfig.
```

camilamacedo86 commented:
@joelanford it also needs to update the .website/

asmacdo commented on doc/sdk-cli-reference.md:
```diff
@@ -493,6 +493,8 @@ Runs the tests locally
 * `--local-operator-flags` string - flags that the operator needs, while using --up-local (e.g. \"--flag1 value1 --flag2=value2\")
 * `--no-setup` - disable test resource creation
 * `--image` string - use a different operator image from the one specified in the namespaced manifest
+* --operator-namespace string     Namespace where the operator will be deployed, CRs will be created and tests will be executed (By default it will be in the default namespace defined in the kubeconfig)
```
FYI this file is about to be removed. https://github.com/operator-framework/operator-sdk/pull/2682

nikhil-thomas replied:
copy that. :+1:
 
shall i leave it there, or do you want me to remove it.

nikhil-thomas replied:
@asmacdo i have rebased again.
